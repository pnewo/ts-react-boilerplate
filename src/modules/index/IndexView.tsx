import * as React from 'react';
import Todo from '../../common/Todo';
import TodoComponent from '../../components/TodoComponent';
import Button from '../../components/Button';
import Loader from '../../components/Loader';

export interface IIndexState {
    title: string;
    todos: Todo[];
    loading: boolean;
}

export interface IIndexDispatch {
    setTitle(n: string): void;
    saveTodo(): void;
    setDone(i: number): void;
}

export type IIndexProps = IIndexState & IIndexDispatch;

const IndexView: React.StatelessComponent<IIndexProps> = ({ title, todos, loading, setTitle, saveTodo, setDone }) => (
    <main className="index">
        {loading && <Loader />}
        <h1 className="index__header">Todo app</h1>
        <form className="index__form" onSubmit={e => e.preventDefault()}>
            <label className="index__form__label" htmlFor="newtodo">Add a new todo:</label>
            <input
                className="index__form__input"
                name="newtodo"
                type="text"
                autoFocus={true}
                value={title}
                onChange={e => setTitle(e.target.value)}
            />
            <Button click={saveTodo} text="Add" />
        </form>
        <br />
        <section className="index__todo-container">
            {todos.map(t => <TodoComponent todo={t} setDone={setDone} key={t.id} />)}
        </section>
    </main>
);

export default IndexView;
