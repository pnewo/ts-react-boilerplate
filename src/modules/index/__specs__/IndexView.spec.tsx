import * as React from 'react';
import { shallow } from 'enzyme';
import IndexView from '../IndexView';
import Todo from '../../../common/Todo';

describe('IndexView', () => {
    const testTodo1 = new Todo(0, 'title');
    const testTodo2 = new Todo(1, 'testing', true);
    const testTitle = 'A title';
    const testSetTitle = jest.fn();
    const testSaveTodo = jest.fn();
    const testSetDone = jest.fn();
    const wrapperMinimalProps = shallow((
        <IndexView
            title=""
            todos={[]}
            loading={false}
            setTitle={testSetTitle}
            saveTodo={testSaveTodo}
            setDone={testSetDone}
        />
    ));
    const wrapperMaximumProps = shallow((
        <IndexView
            title={testTitle}
            todos={[testTodo1, testTodo2]}
            loading={true}
            setTitle={testSetTitle}
            saveTodo={testSaveTodo}
            setDone={testSetDone}
        />
    ));

    it('should render with correct props', () => {
        expect(wrapperMinimalProps).toMatchSnapshot();
        expect(wrapperMaximumProps).toMatchSnapshot();
    });

    it('should call the correct functions when typing to input field', () => {
        const testValue = 'A_TEST_VALUE';
        wrapperMinimalProps.find('[type="text"]').simulate('change', { target: { value: testValue }});
        expect(testSetTitle).toBeCalledWith(testValue);
    });
});
