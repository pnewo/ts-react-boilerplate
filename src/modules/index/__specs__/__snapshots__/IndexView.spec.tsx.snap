// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexView should render with correct props 1`] = `
<main
  className="index"
>
  <h1
    className="index__header"
  >
    Todo app
  </h1>
  <form
    className="index__form"
    onSubmit={[Function]}
  >
    <label
      className="index__form__label"
      htmlFor="newtodo"
    >
      Add a new todo:
    </label>
    <input
      autoFocus={true}
      className="index__form__input"
      name="newtodo"
      onChange={[Function]}
      type="text"
      value=""
    />
    <Button
      click={[Function]}
      text="Add"
    />
  </form>
  <br />
  <section
    className="index__todo-container"
  />
</main>
`;

exports[`IndexView should render with correct props 2`] = `
<main
  className="index"
>
  <Loader />
  <h1
    className="index__header"
  >
    Todo app
  </h1>
  <form
    className="index__form"
    onSubmit={[Function]}
  >
    <label
      className="index__form__label"
      htmlFor="newtodo"
    >
      Add a new todo:
    </label>
    <input
      autoFocus={true}
      className="index__form__input"
      name="newtodo"
      onChange={[Function]}
      type="text"
      value="A title"
    />
    <Button
      click={[Function]}
      text="Add"
    />
  </form>
  <br />
  <section
    className="index__todo-container"
  >
    <TodoComponent
      setDone={[Function]}
      todo={
        Todo {
          "done": false,
          "id": 0,
          "setDone": [Function],
          "title": "title",
        }
      }
    />
    <TodoComponent
      setDone={[Function]}
      todo={
        Todo {
          "done": true,
          "id": 1,
          "setDone": [Function],
          "title": "testing",
        }
      }
    />
  </section>
</main>
`;
